File: src/Tests/src/SimpleTests.cpp
#include <boost/filesystem.hpp>

#include "GTestGroups.h"

#include <Schema/simple.pa.h>

    using namespace boost;
    using namespace bsw;
    using namespace protoargs;

    class SimpleTests: public LogTest
    {
        public:
            SimpleTests() {}
            virtual ~SimpleTests() {}

        public:
            virtual void SetUp() override
            {
            }
    };

    TEST_F(SimpleTests, SimpleUsage)
    {
       const char* argv[] = {
          "program"
          ,"--help"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       simple::ProtoArgs arguments;
       auto config = std::shared_ptr<simple::protoargs>( arguments.parse(argv[0], argc, (char**)argv, true /*allow incomplete*/) );
       ASSERT_TRUE(config != nullptr);

       ASSERT_TRUE(config->help());

       const auto usage = arguments.usage(argv[0]);
       printMessage( usage );
       ASSERT_FALSE( usage.empty() );
    }

    TEST_F(SimpleTests, CheckAllPositive)
    {
       const char* argv[] = {
          "program"
          , "--count", "1"
          , "--configuration", "/tmp/conf"
          , "--flags", "true"
          , "--flags", "false"
          , "-c", "flags should be true and false"
          , "--o-underscore", "no underscore"
          , "--r-underscore", "no underscore"
          , "--a-underscore", "no underscore0"
          , "--a-underscore", "no underscore1"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       simple::ProtoArgs arguments;
       auto config = std::shared_ptr<simple::protoargs>( arguments.parse(argv[0], argc, (char**)argv) );
       ASSERT_TRUE(config != nullptr);

       ASSERT_EQ( 1u, config->count() );

       ASSERT_EQ( "/tmp/conf", config->configuration() );

       ASSERT_EQ( 2, config->flags_size() );
       ASSERT_TRUE( config->flags(0) );
       ASSERT_FALSE( config->flags(1) );

       ASSERT_EQ( "flags should be true and false", config->c() );

       ASSERT_EQ( "no underscore", config->r_underscore() );
       ASSERT_EQ( "no underscore", config->o_underscore() );

       ASSERT_EQ( 2, config->a_underscore_size() );
       ASSERT_EQ( "no underscore0", config->a_underscore(0) );
       ASSERT_EQ( "no underscore1", config->a_underscore(1) );

    }

    // positional will be skipped
    TEST_F(SimpleTests, CheckNoPositional)
    {
       const char* argv[] = {
          "program"
          , "--count", "1"
          , "--configuration", "/tmp/conf"
          , "--flags", "true"
          , "--flags", "false"
          , "-c", "flags should be true and false"
          , "--r-underscore", "no underscore"
          , "positional_value"
          , "positional_value"
          , "positional_value"
          , "positional_value"
          , "positional_value"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       simple::ProtoArgs arguments;
       auto config = std::shared_ptr<simple::protoargs>( arguments.parse(argv[0], argc, (char**)argv) );
       ASSERT_TRUE(config != nullptr);

       ASSERT_EQ( 1u, config->count() );

       ASSERT_EQ( "/tmp/conf", config->configuration() );

       ASSERT_EQ( 2, config->flags_size() );
       ASSERT_TRUE( config->flags(0) );
       ASSERT_FALSE( config->flags(1) );

       ASSERT_EQ( "flags should be true and false", config->c() );
    }

    TEST_F(SimpleTests, CheckWrongType)
    {
       const char* argv[] = {
          "program"
          , "--count", "stringnotdigit"
          , "--configuration", "/tmp/conf"
          , "--flags", "true"
          , "--flags", "false"
          , "-c", "flags should be true and false"
       };
       int argc = sizeof(argv)/sizeof(argv[0]);
       simple::ProtoArgs arguments;
       auto config = std::shared_ptr<simple::protoargs>( arguments.parse(argv[0], argc, (char**)argv) );
       ASSERT_TRUE(config == nullptr);
    }

File: src/TestsRust/src/simple_tests.rs
//use crate::simple_pa;
//#[path = "simple_pa.rs"] mod simple_pa;

#[cfg(test)]
mod simple_tests {

    // Note this useful idiom: importing names from outer (for mod tests) scope.
    //use super::*;

    #[test]
    fn simple_usage() {
        let description = "Desription";
        let argv = vec!["program", "--help"];

        // check usage
        let usage = crate::simple_pa::simple_pa::usage(argv[0], description);
        assert_eq!(usage.trim().is_empty(), false);
    }

    #[test]
    fn check_all_positive() {
        let description = "Desription";
        let argv = vec![ "program"
           , "--count", "1"
           , "--configuration", "/tmp/conf"
           , "--flags", "true"
           , "--flags", "false"
           , "-c", "flags should be true and false"
           , "--o-underscore", "no underscore"
           , "--r-underscore", "no underscore"
           , "--a-underscore", "no underscore0"
           , "--a-underscore", "no underscore1"
        ];

        let rconfig = crate::simple_pa::simple_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( true, rconfig.is_ok(), "{}", rconfig.err().unwrap() );
        let config = rconfig.ok().unwrap();

        assert_eq!( 1, config.count() );

        assert_eq!( "/tmp/conf", config.configuration() );

        assert_eq!( 2, config.flags().len() );
        assert_eq!( true, config.get_flags(0) );
        assert_eq!( false, config.get_flags(1) );

        assert_eq!( "flags should be true and false", config.c() );

        assert_eq!( "no underscore", config.r_underscore() );
        assert_eq!( "no underscore", config.o_underscore() );

        assert_eq!( 2, config.a_underscore().len() );
        assert_eq!( "no underscore0", config.get_a_underscore(0) );
        assert_eq!( "no underscore1", config.get_a_underscore(1) );
    }

    // trailing positional will be skipped
    #[test]
    //#[ignore] // FIXME clap fails on trailing positionals if they are not covered with config and
    // FIXME do not continue other tests
    fn check_no_positional() {
        let description = "Desription";
        let argv = vec![ "program"
           , "--count", "1"
           , "--configuration", "/tmp/conf"
           , "--flags", "true"
           , "--flags", "false"
           , "-c", "flags should be true and false"
           , "--r-underscore", "no underscore"
           , "positional_value"
           , "positional_value"
           , "positional_value"
           , "positional_value"
           , "positional_value"
        ];

        let rconfig = crate::simple_pa::simple_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( true, rconfig.is_ok(), "{}", rconfig.err().unwrap() );
        let config = rconfig.ok().unwrap();

        assert_eq!( 1, config.count() );

        assert_eq!( "/tmp/conf", config.configuration() );

        assert_eq!( 2, config.flags().len() );
        assert_eq!( true, config.get_flags(0) );
        assert_eq!( false, config.get_flags(1) );

        assert_eq!( "flags should be true and false", config.c() );
    }


    #[test]
    fn check_wrong_type() {
        let description = "Desription";
        let argv = vec![ "program"
           , "--count", "stringnotdigit"
           , "--configuration", "/tmp/conf"
           , "--flags", "true"
           , "--flags", "false"
           , "-c", "flags should be true and false"
           , "--r-underscore", "no underscore"
        ];

        let rconfig = crate::simple_pa::simple_pa::parse_debug(argv[0], &argv[..], description);
        assert_eq!( true, rconfig.is_err() );
    }
}
